runtime/unix.c

```diff
@@ -475,6 +478,108 @@ int caml_num_rows_fd(int fd)
 
 void caml_init_os_params(void)
 {
-  caml_sys_pagesize = sysconf(_SC_PAGESIZE);
+  caml_plat_mmap_alignment = caml_plat_pagesize = sysconf(_SC_PAGESIZE);
   return;
 }
+
+#ifndef __CYGWIN__
+
+/* Standard Unix implementation: reserve with mmap (and trim to alignment) with
+   commit done using mmap as well. */
+
+void *caml_plat_mem_map(uintnat size, uintnat alignment, int reserve_only)
+{
+  uintnat alloc_sz = size + alignment;
+  uintnat base, aligned_start, aligned_end;
+  void* mem;
+
+  mem = mmap(0, alloc_sz, reserve_only ? PROT_NONE : (PROT_READ | PROT_WRITE),
+             MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+  if (mem == MAP_FAILED)
+    return 0;
+
+  /* trim to an aligned region */
+  base = (uintnat)mem;
+  aligned_start = (base + alignment - 1) & ~(alignment - 1);
+  aligned_end = aligned_start + size;
+  caml_gc_message(0x1000, "munmap %" ARCH_INTNAT_PRINTF_FORMAT "d"
+                          " bytes at %p for heaps\n",
+                          aligned_start - base, (void*)base);
+  munmap(mem, aligned_start - base);
+  caml_gc_message(0x1000, "munmap %" ARCH_INTNAT_PRINTF_FORMAT "d"
+                          " bytes at %p for heaps\n",
+                          (base + alloc_sz) - aligned_end, (void*)aligned_end);
+  munmap((void*)aligned_end, (base + alloc_sz) - aligned_end);
```

Comment:
You should avoid calling `munmap` with length 0, i.e. if the address returned by `mmap` is suitably aligned already.  Not only this saves on system calls, but some standards say that `munmap` should fail if the length parameter is 0: https://pubs.opengroup.org/onlinepubs/9699919799/functions/munmap.html

Reply:
Thanks for the review! The code in 5.0 is definitely calling `munmap` with length `0`, so a version of this should be extracted to go on 5.0. Would you prefer that I do a minimal patch, guarding the existing calls to `munmap`, or extract the fully split implementation from this PR? (@Octachron deferred to you 🙂)

Reply:
For 5.0, I think we can either do nothing (hoping that all reasonable implementations of `munmap` accept zero lengths) or do a quick fix with a guard
```
   munmap(addr, len)    --->  if (len != 0) munmap(addr, len);
```
A backport from this PR feels too risky at this late point in the 5.0 release cycle.